answers.md

```diff
@@ -0,0 +1,35 @@
+#### Questions:
+1. When this code is run in Node, e.g. `node index.js`, what are the two stages of execution for this file called, and which order do they happen in?
+
+- The two stages are compilation and execution, and they occur in that order.
+
+2. Write an explanation, using as much space as you need, relating to how the first stage of execution for this file operates.
+    - For example, identify the high level steps in a line by line overview and then define what each of those steps are accomplishing.
+
+- During the compilation phase, all variable and function declarations are hoisted and saved in memory. We have not assigned values to the variables as that part comes during the execution phase. foo and bar() are both assigned in memory on the global scope. Within the bar() function, foo is re-declared within the function scope, as well as the function baz() gets declared. foo is once again re-declared within the baz function.
+
+3. Write an explanation, using as much space as you need, relating to how the second stage of execution for this file operates.
+    - For example, identify the high level steps in a line by line overview and then define what each of those steps are accomplishing.
+
+- The second stage of execution will actually assign the values to the declared variables in the code. In the global scope, foo is assigned the value of 'bar'. Within the bar() function, foo is re-assigned 'baz', and is again re-assigned 'bam' within the scope of the baz() function. It attempts to assign 'yay' to the variable bam within the baz function, however since bam was never actually declared, this will cause an error. Functions will also be executed during this phase.
+
+4. During the second stage of execution how many scopes have been registered by the engine?
+    - Which segments of the code do they belong to?
+    - Please identify any variables/refs and which scope each belongs to?
+
+- Three scopes have been registered by the engine: The global scope, the scope within function bar(), and the scope within function baz(). Working outward, baz()'s scope extends from lines 8-12, bar()'s from 5-14, and the global is everything outside of bar() & baz()'s scope. The lines that include baz()'s scope are not included in bar()'s scope. The foo variable belongs to every scope, as it is assigned to 'bar' in the global scope, 'baz' in the bar() scope, and 'bam' in the baz() scope. The other variable assignment, bam is in the baz() scope, however it is not defined where and therefore will throw an error. The function bar() is in the global scope and function baz() within the bar() scope.
```

Comment:
Again, `foo` is actually operating a little differently than expressed here. Line 3 foo belongs to global scope. Line 6 foo belongs to bar. Line 10 foo travels up the scope chain to find the nearest declared `foo`, where it stops at Line 10 above in the upper `bar` scope and reassigns it.